web/src/saved_replies.ts

```diff
@@ -0,0 +1,161 @@
+import $ from "jquery";
+import assert from "minimalistic-assert";
+import type * as tippy from "tippy.js";
+
+import render_add_saved_reply_modal from "../templates/add_saved_reply_modal.hbs";
+
+import * as channel from "./channel";
+import * as compose_ui from "./compose_ui";
+import * as dialog_widget from "./dialog_widget";
+import * as dropdown_widget from "./dropdown_widget";
+import type {Option} from "./dropdown_widget";
+import {$t, $t_html} from "./i18n";
+import * as keydown_util from "./keydown_util";
+import type {StateData} from "./state_data";
+import * as util from "./util";
+
+export type SavedReplies = {
+    id: number;
+    title: string;
+    content: string;
+    rendered_content: string;
+    date_created: number;
+};
+
+const ADD_SAVED_REPLY_OPTION_ID = -1;
+let my_saved_replies: SavedReplies[] = [];
+
+export function set_saved_replies(saved_replies: SavedReplies[]): void {
+    my_saved_replies = saved_replies;
+    my_saved_replies.sort((a, b) => util.strcmp(a.title.toLowerCase(), b.title.toLowerCase()));
+}
+
+export function get_saved_replies(): SavedReplies[] {
+    return my_saved_replies;
+}
+
+export function add_saved_reply(saved_reply: SavedReplies): void {
+    my_saved_replies.push(saved_reply);
+    my_saved_replies.sort((a, b) => util.strcmp(a.title.toLowerCase(), b.title.toLowerCase()));
+}
+
+export function remove_saved_reply(saved_reply_id: number): void {
+    my_saved_replies = my_saved_replies.filter((s) => s.id !== saved_reply_id);
+}
```

Comment:
Our standard convention these days is to put the above block of code into `saved_replies.ts`, and the UI code below it into a new `saved_replies_ui.ts`; this separation often ends up being helpful for avoiding import cycles. (The former module imports almost nothing, yet needs to be imported by code paths like `server_events_dispatch.ts`)

Reply:
Actually `get_options_for_dropdown_widget` could probably go into the data module side of that divide too, since it's a nice, pure, testable function.